packages/node/src/integrations/localvariables.ts

```diff
@@ -41,86 +66,109 @@ class AsyncSession implements DebugSession {
   }
 
   /** @inheritdoc */
-  public async getLocalVariables(objectId: string): Promise<Record<string, unknown>> {
-    const props = await this._getProperties(objectId);
-    const unrolled: Record<string, unknown> = {};
-
-    for (const prop of props) {
-      if (prop?.value?.objectId && prop?.value.className === 'Array') {
-        unrolled[prop.name] = await this._unrollArray(prop.value.objectId);
-      } else if (prop?.value?.objectId && prop?.value?.className === 'Object') {
-        unrolled[prop.name] = await this._unrollObject(prop.value.objectId);
-      } else if (prop?.value?.value || prop?.value?.description) {
-        unrolled[prop.name] = prop.value.value || `<${prop.value.description}>`;
+  public getLocalVariables(objectId: string, complete: (vars: Variables) => void): void {
+    this._getProperties(objectId, props => {
+      const { add, next } = createCallbackList<Variables>(complete);
+
+      for (const prop of props) {
+        if (prop?.value?.objectId && prop?.value.className === 'Array') {
+          const id = prop.value.objectId;
+          add(vars => this._unrollArray(id, prop.name, vars, next));
+        } else if (prop?.value?.objectId && prop?.value?.className === 'Object') {
+          const id = prop.value.objectId;
+          add(vars => this._unrollObject(id, prop.name, vars, next));
+        } else if (prop?.value?.value || prop?.value?.description) {
+          add(vars => this._unrollOther(prop, vars, next));
+        }
       }
-    }
 
-    return unrolled;
+      next({});
+    });
   }
 
   /**
    * Gets all the PropertyDescriptors of an object
    */
-  private _getProperties(objectId: string): Promise<Runtime.PropertyDescriptor[]> {
-    return new Promise((resolve, reject) => {
-      this._session.post(
-        'Runtime.getProperties',
-        {
-          objectId,
-          ownProperties: true,
-        },
-        (err, params) => {
-          if (err) {
-            reject(err);
-          } else {
-            resolve(params.result);
-          }
-        },
-      );
-    });
+  private _getProperties(objectId: string, complete: (result: Runtime.PropertyDescriptor[]) => void): void {
+    this._session.post(
+      'Runtime.getProperties',
+      {
+        objectId,
+        ownProperties: true,
+      },
+      (err, params) => {
+        if (err) {
+          complete([]);
+        } else {
+          complete(params.result);
+        }
+      },
+    );
   }
 
   /**
    * Unrolls an array property
    */
-  private async _unrollArray(objectId: string): Promise<unknown> {
-    const props = await this._getProperties(objectId);
-    return props
-      .filter(v => v.name !== 'length' && !isNaN(parseInt(v.name, 10)))
-      .sort((a, b) => parseInt(a.name, 10) - parseInt(b.name, 10))
-      .map(v => v?.value?.value);
+  private _unrollArray(objectId: string, name: string, vars: Variables, complete: (vars: Variables) => void): void {
```

Comment:
Ditto with the other private methods 

```suggestion
  private _unrollArray(objectId: string, name: string, vars: Variables, next: (vars: Variables) => void): void {
```